spi: SAM support SPI transfers with DMA

teburd opened the pull request:
For larger transfers DMA can be used enabling other tasks
to continue running. A threshold of 32 byte transfers
is about right and is the set value for using DMA.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
spi: SAM support SPI transfers with DMA

For larger transfers DMA can be used enabling other tasks
to continue running. A threshold of 32 byte transfers
is about right and is defined threshold value for using DMA.

This does not currently support multiple SPI transactions changing
chip select with DMA (though the hardware supports this) currently.
Instead opting for the simpler first change of enabling one shot
DMA SPI transfers for those where the size warrants it.

Adds the loopback binding option to enable the spi_loopback test.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
Draft as I'd like some thoughts on the approach before converting to a normal PR, @nandojve I'm especially interested to hear if you have some thoughts on a better approach here that you might like to see.

nandojve commented:
Hi @teburd ,

I'll try to look on this soon. This could be interesting use case for IEEE 802.15.4 with RF2xx if #49775 is addressed.

teburd commented:
I have a sample I'm building up using the tdk robokit1 and the with its 42688-p sensor using this, and I can safely say its working just fine now.

nandojve commented on dts/bindings/spi/atmel,sam-spi.yaml:
```diff
@@ -20,3 +20,21 @@ properties:
+      required: false
+      description: |
+        TX & RX dma specifiers.  Each specifier will have a phandle
+        reference to the dma controller, the channel number, and peripheral
+        trigger source.
```
You need clarify how user will select the channel number and peripheral trigger. It must be clear how user select this options and where he can find the information. This info should be 100% clear without using datasheet as devicetree maintainers have been requesting.

In the case of channel number [0..23], is there a dedicated channel for each peripheral or it can be any arbitrary number? In the case of arbitrary number, is there priority involved?

nandojve reviewed (changes requested):
Hi @teburd ,

I noted that the current PR not implement all details. In this case, I think may be important mention what is possible achieve using current DMA version on commit message. It could be good mention that CSAAT/CSNAAT and LASTXFER are out of scope for now and what is the default behavior in regards of CS selection/deselection.

I think SPI/DMA may not have been build on CI and it is important have a use case to avoid regressions.

nandojve commented:
Hi @teburd ,

> I think SPI/DMA may not have been build on CI and it is important have a use case to avoid regressions.

Do you know if there is a test ? If there is I think we can move this forward.

teburd commented:
The test case is sort of a work in progress, I could add the channel DTS info though so perhaps at least when building the spi tests it would build this with xdmac usage. Let me try that

Commit:
boards: tdk_robokit1: Enable SPI0 DMA usage

SPI0 is connected to the TDK 42688-p sensor which produces data
at a very fast rate. Enable using DMA with this SPI bus
by default for this board.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
@nandojve In order to ensure the driver is built with DMA I needed to enable DMA by default for the board. This seems pretty reasonable to me, without it the SPI reads of this sensor can take up significant CPU cycles spinning.

nandojve commented on boards/arm/tdk_robokit1/tdk_robokit1-common.dtsi:
```diff
@@ -81,7 +81,8 @@
 	pinctrl-0 = <&spi0_default>;
 	pinctrl-names = "default";
-
+	dmas = <&xdmac 0 DMA_PERID_SPI0_TX>, <&xdmac 1 DMA_PERID_SPI0_RX>;
+	dma-names = "tx", "rx";
```
This is OK but maybe it is necessary add an alias for dma as https://github.com/zephyrproject-rtos/zephyr/blob/78809549ee4cf7de3150ff4c3887011825584547/tests/drivers/dma/scatter_gather/boards/frdm_k64f.overlay#L7-L11

and activate DMA function:
```
&dmac {
	status = "okay";
};
```

teburd replied:
I'm not sure I see the need to alias the DMA here, the board uses a specific dma in this case

nandojve reviewed (changes requested):
Hi @teburd ,

I noted is missing SPI_ASYNC API guards and made suggestions about how to proper enable tests for DMA and SPI+DMA.

With all those entries I think it will be possible to keep driver compatibility between all SAM families, since there are 3 DMA IPs and only 1 is currently implemented.

teburd commented:
> Hi @teburd ,
> 
> I noted is missing SPI_ASYNC API guards and made suggestions about how to proper enable tests for DMA and SPI+DMA.
> 
> With all those entries I think it will be possible to keep driver compatibility between all SAM families, since there are 3 DMA IPs and only 1 is currently implemented.

This adds DMA support to the blocking API. The async API still isn't implemented here as I don't need it.

I'm not sure many of the comments above then make sense.

I had to add CONFIG_DMA=y but maybe I can sort out some Kconfigs to select DMA if there'a  DT attribute setup to use DMA with SPI. Will look around for that.

teburd commented:
Is it possible to accept this as it stands today? I don't think its unreasonable and can easily be improved on if others are looking for improvements on it. What would be the minimal set of changes to get it in?

In looking at what I'd need to have a "select DMA if ..." type statement in spi/Kconfig.sam I'd need something like dt_any_compat_has_prop() (thanks @galak) and that doesn't exist at the moment. It seems like that would be growing the scope of what I'm trying to do in this PR.

teburd commented on boards/arm/tdk_robokit1/tdk_robokit1_defconfig:
```diff
@@ -15,3 +15,6 @@ CONFIG_BUILD_OUTPUT_HEX=y
 
 # Enable HW stack protection
 CONFIG_HW_STACK_PROTECTION=y
+
+# Enable DMA
+CONFIG_DMA=y
```
I tried setting xdmac status = "okay" and DMA wasn't compiled, I needed to explicitly set the DMA config to get this to work.

teburd replied:
Done

nandojve reviewed (changes requested):
Hi @teburd ,

> Is it possible to accept this as it stands today? I don't think its unreasonable and can easily be improved on if others are looking for improvements on it. What would be the minimal set of changes to get it in?

Atmel is driven by community. There is very few people (one, two maybe three) that look on it. In this case, It is important make sure thinks are OK and running on CI. Otherwise, it is almost impossible to keep minimal quality.

As I see, below as mandatory:

* DMA code should be present only when DMA is active. I changed my suggestion to guard by DMA instead SPI_ASYNC, as you pointed that DMA doesn't have nothing to do with ASYNC.
* Add loopback property on devicetree and handle it on driver.
* It is necessary run `tests/drivers/spi/spi_loopback` on CI. To do it, add a config that enables loopback mode with the property that you should introduce.
* rename `spi_sam_xdmac_txrx` to `spi_sam_dma_txrx`.

> In looking at what I'd need to have a "select DMA if ..." type statement in spi/Kconfig.sam I'd need something like dt_any_compat_has_prop() (thanks @galak) and that doesn't exist at the moment. It seems like that would be growing the scope of what I'm trying to do in this PR.

I made a comment that should fix the issue. You can just `select DMA` at Kconfig.sam_xdmac.

Commit:
tests: spi_loopback: Enable SPI testing on SAM

Adds the needed overlays and board dtsi modifications to run the
spi loopback test on the sam e70 xplained, tdk robokit1, and sam v71 xult.

Tests both DMA and non-DMA configurations.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented on drivers/spi/spi_sam.c:
```diff
@@ -215,6 +226,145 @@ static void spi_sam_fast_rx(Spi *regs, const struct spi_buf *rx_buf)
+}
+
+/* DMA path does a transceive when given a large rx buf of significant length (> 16 bytes) */
+static int spi_sam_xdmac_txrx(const struct device *dev, Spi *regs, const struct spi_buf *tx_buf,
```
Done, and I've moved the ifdefry into spi_sam_[tx|rx|txrx] inlines that make the looping over the spi bufs look nicer

teburd commented:
> Hi @teburd ,
> 
> > Is it possible to accept this as it stands today? I don't think its unreasonable and can easily be improved on if others are looking for improvements on it. What would be the minimal set of changes to get it in?
> 
> Atmel is driven by community. There is very few people (one, two maybe three) that look on it. In this case, It is important make sure thinks are OK and running on CI. Otherwise, it is almost impossible to keep minimal quality.
> 
> As I see, below as mandatory:
> 
>     * DMA code should be present only when DMA is active. I changed my suggestion to guard by DMA instead SPI_ASYNC, as you pointed that DMA doesn't have nothing to do with ASYNC.
> 
>     * Add loopback property on devicetree and handle it on driver.
> 
>     * It is necessary run `tests/drivers/spi/spi_loopback` on CI. To do it, add a config that enables loopback mode with the property that you should introduce.
> 
>     * rename `spi_sam_xdmac_txrx` to `spi_sam_dma_txrx`.
> 
> 
> > In looking at what I'd need to have a "select DMA if ..." type statement in spi/Kconfig.sam I'd need something like dt_any_compat_has_prop() (thanks @galak) and that doesn't exist at the moment. It seems like that would be growing the scope of what I'm trying to do in this PR.
> 
> I made a comment that should fix the issue. You can just `select DMA` at Kconfig.sam_xdmac.

SPI with loopbacks, with and without DMA are now tested with the spi_loopback test.

All the things on the list are done

teburd commented:
Noticed a missing ifdef around the dma_sem, fixed

nandojve commented:
Hi @teburd ,

I dig around to understand problem with configuration. I'm sending a patch to you that adjusts configs/dts/files to build with/without DMA support, see [sam_spi_plus_dma.txt](https://github.com/zephyrproject-rtos/zephyr/files/9863313/sam_spi_plus_dma.txt).

Please, cherry-pick changes, let's see who CI will react and let's get this merged :-)

Commit:
dts: same70 Disable xdmac by default

Set the status of the DMA controller, xdmac, to disabled. In effect
changing the default status from okay to disabled for all sam e70
based board.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
> Hi @teburd ,
> 
> I dig around to understand problem with configuration. I'm sending a patch to you that adjusts configs/dts/files to build with/without DMA support, see [sam_spi_plus_dma.txt](https://github.com/zephyrproject-rtos/zephyr/files/9863313/sam_spi_plus_dma.txt).
> 
> Please, cherry-pick changes, let's see how CI will react and let's get this merged :-)

looking through the patch I see you disabled by default the xdmac, isn't this potentially a breaking change for downstream users?

Not a big deal to me, but something I noted. I kept that change in a separate commit.

nandojve commented on boards/arm/tdk_robokit1/tdk_robokit1.yaml:
```diff
@@ -9,6 +9,7 @@ toolchain:
   - gnuarmemb
   - xtools
 supported:
+  - dma
```
I missed this, would you mind to add for `sam_e70_xplained` and `sam_v71_xult`.

nandojve commented on drivers/spi/Kconfig.sam:
```diff
@@ -11,3 +11,12 @@ config SPI_SAM
+	bool "SPI SAM DMA Support"
+	select DMA
+	help
+	  Enable the SPI DMA mode using XDMAC for SPI instances
+	  that enable dma channels in their device tree node.
```
The SPI_SAM_DMA is for all DMA controllers, need to keep text generic. So, you need change the message around `XDMAC`.

nandojve reviewed (changes requested):
> Looking through the patch I see you disabled by default the xdmac, isn't this potentially a breaking change for downstream users?
>
> Not a big deal to me, but something I noted. I kept that change in a separate commit.

Yes, it is intentional. The Kconfig+devicetree logic are enabling a lot of stuff automatically and user don't have any glue that DMA is automatically enabled. For me it is an anti-pattern a user add a node on the board to disable DMA.

> I also kept the SPI_SAM_DMA Kconfig, which aligns with how NXP and ST Kconfig options for SPI and DMA work. It also simplifies the decision of when to enable various DMA features in the driver which is nice.

Particularly I don't see too much advantage at first. However, I imagine that many users may have slow devices which  may save flash and DMA channels. So I'm glad to have it!

> In the future that could maybe be automatically selected if a few DT attributes/statuses are setup in a particular fashion but that's something that would be different than ST or NXP currently.

Agreed.

nandojve reviewed (dismissed):
HI @teburd ,

Thank you so much!

pdgendt commented on drivers/spi/spi_sam.c:
```diff
@@ -467,11 +689,22 @@ static const struct spi_driver_api spi_sam_driver_api = {
+	.dma_rx_channel = DT_INST_DMAS_CELL_BY_NAME(n, rx, channel),                               \
+	.dma_rx_perid = DT_INST_DMAS_CELL_BY_NAME(n, rx, perid),
+
+#define SPI_SAM_USE_DMA(inst, name) DT_INST_DMAS_HAS_NAME(n, tx) && IS_ENABLED(CONFIG_SPI_SAM_DMA)
```
The `name` argument isn't used.

pdgendt commented on drivers/spi/spi_sam.c:
```diff
@@ -215,6 +236,150 @@ static void spi_sam_fast_rx(Spi *regs, const struct spi_buf *rx_buf)
+static void dma_tx_callback(const struct device *dma_dev, void *user_data,
+	uint32_t channel, int status)
+{
+	struct k_sem *sem = user_data;
```
Add `ARG_UNUSED(var)` for the unused arguments.

pdgendt commented on drivers/spi/spi_sam.c:
```diff
@@ -215,6 +236,150 @@ static void spi_sam_fast_rx(Spi *regs, const struct spi_buf *rx_buf)
+
+/* DMA path does a transceive when given a large rx buf of significant length (> 16 bytes) */
+static int spi_sam_dma_txrx(const struct device *dev, Spi *regs, const struct spi_buf *tx_buf,
+			      const struct spi_buf *rx_buf)
```
Nit: fix alignment

pdgendt commented on drivers/dma/dma_sam_xdmac.c:
```diff
@@ -209,6 +209,23 @@ static int sam_xdmac_config(const struct device *dev, uint32_t channel,
 
+	uint32_t xdmac_inc_cfg = 0;
+
+	if (cfg->head_block->source_addr_adj ==  DMA_ADDR_ADJ_INCREMENT
```
Nit: extraneous space

teburd commented on drivers/spi/spi_sam.c:
```diff
@@ -215,6 +236,150 @@ static void spi_sam_fast_rx(Spi *regs, const struct spi_buf *rx_buf)
+}
+
+
+/* DMA path does a transceive when given a large rx buf of significant length (> 16 bytes) */
```
Reworded the comment, Kconfig or DTS could be nice improvements but this works well today without needing another knob to tweak. Another option aside from the fixed cutoff or a Kconfig/dts defined one would be a little LUT of transfer size and SPI clock to decide when to use DMA or polling. No need to confuse people with yet another knob to tweak then.

32 bytes is pretty reasonable though as a starting point.

pdgendt commented on drivers/spi/spi_sam.c:
```diff
@@ -215,6 +236,160 @@ static void spi_sam_fast_rx(Spi *regs, const struct spi_buf *rx_buf)
+static uint8_t tx_dummy;
+static uint8_t rx_dummy;
+
+static void dma_tx_callback(const struct device *dma_dev, void *user_data,
+	uint32_t channel, int status)
+{
+	ARG_UNUSED(dma_dev);
+	ARG_UNUSED(channel);
+	ARG_UNUSED(status);
+
+	struct k_sem *sem = user_data;
+
+	k_sem_give(sem);
+}
+
+static void dma_rx_callback(const struct device *dma_dev, void *user_data,
+	uint32_t channel, int status)
+{
+	ARG_UNUSED(dma_dev);
+	ARG_UNUSED(channel);
+	ARG_UNUSED(status);
+
+	struct k_sem *sem = user_data;
+
+	k_sem_give(sem);
+}
```
These are identical, just have a single `dma_callback` function.

pdgendt commented on dts/bindings/spi/atmel,sam-spi.yaml:
```diff
@@ -20,3 +20,29 @@ properties:
+        for testing.
+
+    dmas:
+      required: false
```
No need to add `required: false`

pdgendt commented on dts/bindings/spi/atmel,sam-spi.yaml:
```diff
@@ -20,3 +20,29 @@ properties:
+           dmas = <&xdmac 1 DMA_PERID_SPI0_TX>, <&xdmac 2 DMA_PERID_SPI0_RX>;
+
+    dma-names:
+      required: false
```
Same here

pdgendt commented on drivers/spi/spi_sam.c:
```diff
@@ -467,11 +699,22 @@ static const struct spi_driver_api spi_sam_driver_api = {
+		.regs = (Spi *)DT_INST_REG_ADDR(n),                                                \
+		.periph_id = DT_INST_PROP(n, peripheral_id),                                       \
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),                                         \
+		.loopback = DT_INST_PROP_OR(n, loopback, 0),                                       \
```
A boolean property can be used as is.

```suggestion
		.loopback = DT_INST_PROP(n, loopback),                                             \
```

nandojve replied:
I was out of date regarding booleans. We can go with `DT_PROP`.

pdgendt commented on drivers/spi/spi_sam.c:
```diff
@@ -467,11 +699,22 @@ static const struct spi_driver_api spi_sam_driver_api = {
-		.regs = (Spi *)DT_INST_REG_ADDR(n),			\
-		.periph_id = DT_INST_PROP(n, peripheral_id),		\
-		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),		\
+#define SPI_DMA_INIT(n)                                                                            \
```
All existing whitespace until the `\` were tabs. I'd prefer to keep using tabs here.

pdgendt commented:
@teburd Can you take a look at the latest review comments? Thanks

pdgendt commented on drivers/dma/dma_sam_xdmac.c:
```diff
@@ -209,6 +209,23 @@ static int sam_xdmac_config(const struct device *dev, uint32_t channel,
 	dev_data->dma_channels[channel].data_size = data_size;
 	LOG_DBG("data_size=%d", data_size);
 
+	uint32_t xdmac_inc_cfg = 0;
```
Didn't notice before, but `xdmac_inc_cfg` doesn't appear to be used? Or am I missing something?

teburd replied:
good catch, it should've been used when doing the channel_cfg. Needed for the spi tx/rx from a dummy value where there should be no increment

pdgendt commented on drivers/spi/spi_sam.c:
```diff
@@ -467,24 +687,35 @@ static const struct spi_driver_api spi_sam_driver_api = {
+	.dma_rx_channel = DT_INST_DMAS_CELL_BY_NAME(n, rx, channel),				\
+	.dma_rx_perid = DT_INST_DMAS_CELL_BY_NAME(n, rx, perid),
+
+#define SPI_SAM_USE_DMA(inst) DT_INST_DMAS_HAS_NAME(n, tx) && IS_ENABLED(CONFIG_SPI_SAM_DMA)
```
Apparently this will always result in `0` because `inst != n`

pdgendt replied:
I tried to take fix it in my PR but [CI](https://github.com/zephyrproject-rtos/zephyr/actions/runs/3968156131) doesn't seem to like it..
